hypnotoad/utils/options.py

```diff
@@ -0,0 +1,491 @@
+                self.__allow_access = False
+
+                # Tell the user where the circular definition was
+                index = key_chain.index(key)
+                raise ValueError(
+                    f"Circular definition of default values. At least one of "
+                    f"{key_chain[index:]} must have a definite value"
+                )
+
+            chain_start = False
+            self.__key_chain.append(key)
+
+        self.__values[key] = self.__default_values[key].evaluate_expression(
+            self, name=key
+        )
+
+        if chain_start:
+            # Tidy up temporary member variable
+            del self.__key_chain
+
+        return self.__values[key]
+
+    def __getattr__(self, key):
+        if self.__allow_access:
+            if key in self.__values or key in self.__default_values:
+                return self.__getitem__(key)
+        return super().__getattr__(key)
+
+    def __contains__(self, key):
+        return key in self.__default_values
+
+    def create(self, values={}):
+        """Create an Options instance
+
+        The members of the created Options are defined by this
+        OptionsFactory instance. Any values passed in the values dict are used,
+        and the rest are set from defaults, which can be expressions depending on other
+        members.
+
+        Parameters
+        ----------
+        values : dict, optional
+            Non-default values to be used
+        """
+
+        # do not modify passed-in values
+        values = deepcopy(dict(values))
+
+        # ignore values for keys not in the list of keys defined in the factory
+        for key in list(values.keys()):
+            if key not in self.__default_values:
+                del values[key]
+
+        # check passed-in values
+        self.__values = {
+            key: _checked(value, meta=self.__default_values[key], name=key,)
+            for key, value in values.items()
+        }
+
+        # make a list of the explicitly-set (non-default) values
+        non_default = [k for k in self.__values]
+
+        # Return new Options instance
+        # Now create all other values from default expressions. Default value evalutation
+        # is implemented in the __getitem__ method so we can pass 'self' to the default
+        # expressions.
+        # Set __allow_access to True to allow internal use of __getitem__, which is
+        # forbidden externally
+        self.__allow_access = True
+        for key in self.__default_values:
+            # Note self.__values[key] is updated in self.__getitem__(key)
+            self[key]
+
+        # Create Options to return
+        new_instance = OptionsFactory.Options(
+            self.__values,
+            {key: self.__default_values[key].doc for key in self.__default_values},
+            non_default,
+        )
+
+        self.__allow_access = False
+
+        # Clean up state of the factory
+        del self.__values
```

Comment:
Does this need to be a member-variable in the first place?

Reply:
Alternatives welcome, but the reason I did this is so that `OptionsFactory` can implement `__getitem__()` and `__getattr__()` methods - this is needed so that the default values can be set (in `WithMeta`) with expressions like `lambda options: options.a + options.b`, where the `OptionsFactory` passes itself as the argument to the expressions. So `OptionsFactory.__getitem__(self, key)` needs `__values` without it being passed as an argument.

Reply:
Maybe one alternative would be to have some kind of dedicated internal class for a temporary mutable object that could be given the `__default_values` and `values` and go through evaluating everything, rather than using the `OptionsFactory` object itself for that?